### PR TITLE
feat(otel): centralize GenAI semantic keys

### LINKED ISSUE #13523: [P2] OpenTelemetry GenAI semantic convention key registry
## Problem

MASC already emits some `gen_ai.*` attributes for keeper agent spans and tool spans, but the keys are duplicated as ad-hoc strings across OTel modules. OpenTelemetry GenAI semantic conventions are still in Development, so scattered strings make future compatibility updates risky and make it unclear which keys are official semconv keys vs MASC-specific extensions.

## First slice

- Centralize current GenAI semantic convention keys in `Otel_genai`.
- Separate official keys from MASC extension keys.
- Update keeper/tool span helpers to use the registry instead of local string literals.
- Add regression tests that lock the emitted names and extension boundary.

## Evidence

[근거] OpenTelemetry GenAI semantic conventions page, checked 2026-05-06 KST, confidence High: https://opentelemetry.io/docs/specs/semconv/gen-ai/
[근거] OpenTelemetry GenAI metrics page lists required attributes including `gen_ai.operation.name` and `gen_ai.provider.name`, checked 2026-05-06 KST, confidence High: https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/

## Scope boundary

This is not a full telemetry rename. Existing legacy `keeper.*` and `tool.*` attributes must continue to be emitted for current dashboards.

### PR DESCRIPTION
**Pull request by jeong-sik**
## Summary

- Centralizes OpenTelemetry GenAI attribute keys in `Otel_genai.Attr_key`.
- Separates official `gen_ai.*` keys, MASC extension keys, and legacy dashboard keys.
- Reuses the registry from the tool dispatch span hook instead of duplicating `gen_ai.*` string literals.
- Extends `test_otel_genai` to lock the key boundaries and tool execution attrs.

Fixes #13523.

## Evidence

[근거] OpenTelemetry GenAI semantic conventions are marked Development and describe the stability opt-in transition plan. Source: https://opentelemetry.io/docs/specs/semconv/gen-ai/ checked 2026-05-06 KST, confidence High.
[근거] OpenTelemetry GenAI metrics list `gen_ai.operation.name` and `gen_ai.provider.name` as required Development attributes and list `invoke_agent` / `execute_tool` as well-known operation values. Source: https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/ checked 2026-05-06 KST, confidence High.

## Validation

- `env DUNE_CACHE=disabled scripts/dune-local.sh build test/test_otel_genai.exe`
- `./_build/default/test/test_otel_genai.exe`
- `git diff --check HEAD~1 HEAD`

### CODE REVIEW

File: lib/otel/otel_dispatch_hook.ml [hunk 1]
```diff
@@ -46,19 +46,19 @@ let on_tool_result (result : Tool_result.t) : Tool_result.t =
...
-       those belong on the parent agent / chat span (Step 2 of #7461),
-       not on the inner tool span which is provider-agnostic. *)
+       provider/model keys are intentionally omitted — those belong on the
+       parent agent / model span, not on the inner tool span which is
+       provider-agnostic. *)
     let gen_ai_attrs =
-      [ ("gen_ai.operation.name", `String "execute_tool");
-        ("gen_ai.tool.name", `String result.tool_name) ]
+      Otel_genai.tool_execution_attrs ~tool_name:result.tool_name
+      |> List.map (fun (key, value) -> key, (value :> OT.value))
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/otel/otel_dispatch_hook.ml]: 동의 — post-hook span payload regression test 추가 필요. mock OT collector 주입 후 actual span 발사 검증. helper unit test와 별개.

File: lib/otel/otel_genai.ml [hunk 2]
```diff
@@ -1,11 +1,69 @@
...
+  let keeper_current_task_id = "keeper.current_task_id"
+  let tool_name = "tool.name"
+  let tool_success = "tool.success"
+  let tool_duration_ms = "tool.duration_ms"
+
+  let official_gen_ai =
+    [ gen_ai_operation_name
+    ; gen_ai_provider_name
+    ; gen_ai_agent_name
+    ; gen_ai_agent_id
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/otel/otel_genai.ml:37]: 동의 — classification list spot-check 약함. 전체 enum coverage test (모든 key가 right group에 분류) 권장. classification drift 시 full sweep으로 잡힘.

**Comment by jeong-sik** [comment]
Live MASC task link: added `task-195` to track this OpenTelemetry GenAI registry PR through verification/merge. It replaces stale cancelled runtime task `task-163`, while GitHub issue #13523 remains the issue tracker.

File: lib/otel/otel_genai.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/otel/otel_genai.ml:37]: Addressed in commit `6a22c6f723`: added `test_attr_key_registry_full_coverage`, which iterates every `official_gen_ai`, `masc_extensions`, and `legacy` key and asserts the expected classification boundaries. The same commit also covers the other OTel review note by adding a dispatch post-hook regression: `test_dispatch_hook_emits_tool_span_payload` installs `Otel_dispatch_hook`, runs `Tool_dispatch.run_post_hooks`, captures the emitted span through a test emitter, and asserts the actual span name + GenAI + legacy + status attributes.

Verification: `git diff --check` passed; focused run passed with `MASC_SKIP_OPAM_LOCK=1 MASC_SKIP_PIN_CHECK=1 MASC_DUNE_THROTTLE=0 DUNE_BUILD_DIR=_build_codex_13525 scripts/dune-local.sh exec ./test/test_otel_genai.exe` (7 tests run). Pin guard was bypassed only because the shared local opam switch was being repinned by other worktrees.

File: lib/otel/otel_genai.ml [hunk 3]
```diff
@@ -1,11 +1,69 @@
...
+    ; gen_ai_conversation_id
+    ; gen_ai_tool_name
+    ]
+  ;;
+
+  let masc_extensions = [ masc_gen_ai_keeper_name; masc_gen_ai_cascade_name ]
+
+  let legacy =
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/otel/otel_genai.ml:45]: Follow-up landed in draft PR #13650: added Attr_key.all_known plus completeness/duplicate assertions so the official/MASC/legacy registry partitions are checked against the known key set. Verification is in the PR body: focused test_otel_genai passes on latest main.
